Source/Core/VideoCommon/TextureConversionShader.cpp

```diff
@@ -720,4 +724,526 @@ const char* GenerateEncodingShader(u32 format, APIType ApiType)
+{
+  // Swizzle bits: 00000123 -> 12312312
+  return (v << 5) | (v << 2) | (v >> 1);
+}
+uint Convert4To8(uint v)
+{
+  // Swizzle bits: 00001234 -> 12341234
+  return (v << 4) | v;
+}
+uint Convert5To8(uint v)
+{
+  // Swizzle bits: 00012345 -> 12345123
+  return (v << 3) | (v >> 2);
+}
+uint Convert6To8(uint v)
+{
+  // Swizzle bits: 00123456 -> 12345612
+  return (v << 2) | (v >> 4);
+}
+
+uint GetTiledTexelOffset(uvec2 block_size, uvec2 coords)
+{
+  uvec2 block = coords / block_size;
+  uvec2 offset = coords % block_size;
+  uint buffer_pos = u_src_offset;
+  buffer_pos += block.y * u_src_row_stride;
+  buffer_pos += block.x * (block_size.x * block_size.y);
+  buffer_pos += offset.y * block_size.x;
+  buffer_pos += offset.x;
+  return buffer_pos;
+}
+
+uvec4 GetPaletteColor(uint index)
+{
+  // Fetch and swap BE to LE
+  uint val = texelFetch(s_palette_buffer, int(u_palette_offset + index)).x;
+  val = ((val >> 8) | (val << 8)) & 0xFFFFu;
+
+  uvec4 color;
+#if defined(PALETTE_FORMAT_IA8)
+  uint a = bitfieldExtract(val, 0, 8);
+  uint i = bitfieldExtract(val, 8, 8);
+  color = uvec4(i, i, i, a);
+#elif defined(PALETTE_FORMAT_RGB565)
+  color.x = Convert5To8(bitfieldExtract(val, 11, 5));
+  color.y = Convert6To8(bitfieldExtract(val, 5, 6));
+  color.z = Convert5To8(val & 0x1Fu);
+  color.a = 255u;
+
+#elif defined(PALETTE_FORMAT_RGB5A3)
+  if ((val & 0x8000u) != 0u)
+  {
+    color.x = Convert5To8(bitfieldExtract(val, 10, 5));
+    color.y = Convert5To8(bitfieldExtract(val, 5, 5));
+    color.z = Convert5To8(val & 0x1Fu);
+    color.a = 255u;
+  }
+  else
+  {
+    color.a = Convert3To8(bitfieldExtract(val, 12, 3));
+    color.r = Convert4To8(bitfieldExtract(val, 8, 4));
+    color.g = Convert4To8(bitfieldExtract(val, 4, 4));
+    color.b = Convert4To8(val & 0xFu);
+  }
+#else
+  // Not used.
+  color = uvec4(0, 0, 0, 0);
+#endif
+
+  return color;
+}
+
+vec4 GetPaletteColorNormalized(uint index)
+{
+  uvec4 color = GetPaletteColor(index);
+  return vec4(color) / 255.0;
+}
+
+)";
+
+static const std::array<DecodingShaderInfo, 10> s_decoding_shader_info = {
+    {{GX_TF_I4, BUFFER_FORMAT_R8_UINT, 0, 8, 8, false,
+      R"(
+      layout(local_size_x = 8, local_size_y = 8) in;
+
+      void main()
+      {
+        uvec2 coords = gl_GlobalInvocationID.xy;
+
+        // Tiled in 8x8 blocks, 4 bits per pixel
+        // We need to do the tiling manually here because the texel size is smaller than
+        // the size of the buffer elements.
+        uint2 block = coords.xy / 8u;
+        uint2 offset = coords.xy % 8u;
+        uint buffer_pos = u_src_offset;
+        buffer_pos += block.y * u_src_row_stride;
+        buffer_pos += block.x * 32u;
+        buffer_pos += offset.y * 4u;
+        buffer_pos += offset.x / 2u;
+
+        // Select high nibble for odd texels, low for even.
+        uint val = texelFetch(s_input_buffer, int(buffer_pos)).x;
+        uint i;
+        if ((coords.x & 1u) == 0u)
+          i = Convert4To8((val >> 4));
```

Comment:
As the inputs of Convert4To8 doesn't mask, we need to do it here. At least I guess the input is an unmasked 32 bit value.